File: app/src/main/java/ru/melodin/fast/fragment/FragmentFriends.java
package ru.melodin.fast.fragment;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.widget.PopupMenu;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import java.util.ArrayList;

import ru.melodin.fast.MessagesActivity;
import ru.melodin.fast.R;
import ru.melodin.fast.adapter.UserAdapter;
import ru.melodin.fast.api.UserConfig;
import ru.melodin.fast.api.VKApi;
import ru.melodin.fast.api.model.VKConversation;
import ru.melodin.fast.api.model.VKUser;
import ru.melodin.fast.common.ThemeManager;
import ru.melodin.fast.concurrent.AsyncCallback;
import ru.melodin.fast.concurrent.ThreadExecutor;
import ru.melodin.fast.current.BaseFragment;
import ru.melodin.fast.database.CacheStorage;
import ru.melodin.fast.database.DatabaseHelper;
import ru.melodin.fast.util.ArrayUtil;
import ru.melodin.fast.util.Util;
import ru.melodin.fast.view.FastToolbar;

public class FragmentFriends extends BaseFragment implements SwipeRefreshLayout.OnRefreshListener {

    private static final int FRIENDS_COUNT = 30;

    private FastToolbar tb;
    private View empty;
    private RecyclerView list;
    private SwipeRefreshLayout refreshLayout;

    private UserAdapter adapter;

    private boolean loading;

    public boolean isLoading() {
        return loading;
    }

    public void setLoading(boolean loading) {
        this.loading = loading;
    }

    @Override
    public void onRefresh() {
        getFriends(FRIENDS_COUNT, 0);
    }

    @Override
    public void onDestroy() {
        if (adapter != null) adapter.destroy();
        super.onDestroy();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setTitle(getString(R.string.fragment_friends));
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        initViews(view);

        setToolbar(tb);
        setRecyclerView(list);

        tb.setTitle(getTitle());

        refreshLayout.setOnRefreshListener(this);
        refreshLayout.setColorSchemeColors(ThemeManager.getAccent());
        refreshLayout.setProgressBackgroundColorSchemeColor(ThemeManager.getPrimary());

        LinearLayoutManager manager = new LinearLayoutManager(getActivity(), RecyclerView.VERTICAL, false);
        list.setHasFixedSize(true);

        list.setItemViewCacheSize(20);
        list.setDrawingCacheEnabled(true);
        list.setDrawingCacheQuality(View.DRAWING_CACHE_QUALITY_HIGH);

        list.setLayoutManager(manager);

        getCachedFriends();
        if (savedInstanceState == null)
            getFriends(FRIENDS_COUNT, 0);
    }

    private void initViews(@NonNull View v) {
        tb = v.findViewById(R.id.tb);
        empty = v.findViewById(R.id.no_items_layout);
        list = v.findViewById(R.id.list);
        refreshLayout = v.findViewById(R.id.refresh);
    }

    private void checkCount() {
        empty.setVisibility(adapter == null ? View.VISIBLE : adapter.isEmpty() ? View.VISIBLE : View.GONE);
    }

    private void createAdapter(ArrayList<VKUser> friends, int offset) {
        if (ArrayUtil.isEmpty(friends)) return;

        checkCount();

        if (adapter == null) {
            adapter = new UserAdapter(this, friends);
            list.setAdapter(adapter);

            checkCount();
            return;
        }

        if (offset != 0) {
            adapter.changeItems(friends);
            adapter.notifyDataSetChanged();

            checkCount();
            return;
        }

        adapter.changeItems(friends);
        adapter.notifyDataSetChanged();

        checkCount();
    }

    private void getCachedFriends() {
        ArrayList<VKUser> users = CacheStorage.getFriends(UserConfig.userId, false);

        if (!ArrayUtil.isEmpty(users))
            createAdapter(users, 0);
    }

    private void getFriends(final int count, final int offset) {
        if (loading) return;
        if (!Util.hasConnection()) {
            refreshLayout.setRefreshing(false);
            Toast.makeText(getActivity(), R.string.connect_to_the_internet, Toast.LENGTH_SHORT).show();
            return;
        }

        loading = true;

        refreshLayout.setRefreshing(true);
        ThreadExecutor.execute(new AsyncCallback(getActivity()) {

            private ArrayList<VKUser> users;

            @Override
            public void ready() throws Exception {
                users = VKApi.friends().get().userId(UserConfig.userId).order("hints").fields(VKUser.FIELDS_DEFAULT).execute(VKUser.class);

                if (offset == 0) {
                    CacheStorage.delete(DatabaseHelper.FRIENDS_TABLE);
                    CacheStorage.insert(DatabaseHelper.FRIENDS_TABLE, users);
                }

                CacheStorage.insert(DatabaseHelper.USERS_TABLE, users);
            }

            @Override
            public void done() {
                createAdapter(users, offset);
                loading = true;
                refreshLayout.setRefreshing(false);
            }

            @Override
            public void error(Exception e) {
                refreshLayout.setRefreshing(false);
                Toast.makeText(getActivity(), getString(R.string.error), Toast.LENGTH_LONG).show();
            }
        });
    }

    public void openChat(int position) {
        VKUser user = adapter.getItem(position);

        Intent intent = new Intent(getActivity(), MessagesActivity.class);
        intent.putExtra("title", user.toString());
        intent.putExtra("photo", user.getPhoto200());
        intent.putExtra("peer_id", user.getId());

        boolean canWrite = !user.isDeactivated();

        intent.putExtra("can_write", canWrite);

        if (!canWrite) {
            intent.putExtra("reason", VKConversation.getReason(VKConversation.Reason.USER_DELETED));
        }

        startActivity(intent);
    }

    public void showDialog(final int position, View v) {
        PopupMenu menu = new PopupMenu(getActivity(), v);
        menu.inflate(R.menu.fragment_friends_funcs);
        menu.setOnMenuItemClickListener(item -> {
            showConfirmDeleteFriend(position);
            return true;
        });
        menu.show();
    }

    private void showConfirmDeleteFriend(final int position) {
        AlertDialog.Builder adb = new AlertDialog.Builder(getActivity());
        adb.setTitle(R.string.confirmation);
        adb.setMessage(R.string.confirm_delete_friend);
        adb.setPositiveButton(R.string.yes, (dialog, which) -> deleteFriend(position));
        adb.setNegativeButton(R.string.no, null);
        adb.create().show();
    }

    private void deleteFriend(final int position) {
        if (!Util.hasConnection()) {
            refreshLayout.setRefreshing(false);
            return;
        }

        refreshLayout.setRefreshing(true);

        VKUser user = adapter.getItem(position);
        final int userId = user.getId();

        ThreadExecutor.execute(new AsyncCallback(getActivity()) {

            @Override
            public void ready() throws Exception {
                VKApi.friends().delete().userId(userId).execute();
            }

            @Override
            public void done() {
                adapter.remove(position);
                adapter.notifyItemRemoved(position);
                adapter.notifyItemRangeChanged(0, adapter.getItemCount(), -1);
                refreshLayout.setRefreshing(false);

                CacheStorage.delete(DatabaseHelper.FRIENDS_TABLE, DatabaseHelper.USER_ID, userId);
            }

            @Override
            public void error(Exception e) {
                Log.e("Error delete friend", Log.getStackTraceString(e));
                refreshLayout.setRefreshing(false);
                Toast.makeText(getActivity(), R.string.error, Toast.LENGTH_SHORT).show();
            }
        });
    }
}

File: app/src/main/java/ru/melodin/fast/fragment/FragmentFriends.kt
package ru.melodin.fast.fragment

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.widget.PopupMenu
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import kotlinx.android.synthetic.main.list_empty.*
import kotlinx.android.synthetic.main.recycler_list.*
import kotlinx.android.synthetic.main.toolbar.*
import ru.melodin.fast.R
import ru.melodin.fast.adapter.UserAdapter
import ru.melodin.fast.api.OnResponseListener
import ru.melodin.fast.api.UserConfig
import ru.melodin.fast.api.VKApi
import ru.melodin.fast.api.model.VKConversation
import ru.melodin.fast.api.model.VKUser
import ru.melodin.fast.common.FragmentSelector
import ru.melodin.fast.common.TaskManager
import ru.melodin.fast.common.ThemeManager
import ru.melodin.fast.current.BaseFragment
import ru.melodin.fast.database.CacheStorage
import ru.melodin.fast.database.CacheStorage.getFriends
import ru.melodin.fast.database.DatabaseHelper
import ru.melodin.fast.database.MemoryCache
import ru.melodin.fast.mvp.contract.UsersContract
import ru.melodin.fast.mvp.presenter.UsersPresenter
import ru.melodin.fast.util.ArrayUtil
import ru.melodin.fast.util.Keys
import ru.melodin.fast.util.Util

class FragmentFriends(var onlyOnline: Boolean) : BaseFragment(),
    SwipeRefreshLayout.OnRefreshListener,
    UsersContract.View {

    constructor() : this(false)

    private var adapter: UserAdapter? = null

    var isLoading: Boolean = false

    val presenter = UsersPresenter()

    override fun onRefresh() {
        loadUsers(FRIENDS_COUNT, 0)
    }

    override fun onDestroy() {
        adapter?.destroy()
        super.onDestroy()

        presenter.detachView()
        presenter.destroy()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        presenter.attachView(this)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.fragment, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        toolbar = tb.apply {
            visibility = View.GONE
        }

        recyclerList = list

        refreshLayout.setOnRefreshListener(this)
        refreshLayout.setColorSchemeColors(ThemeManager.ACCENT)
        refreshLayout.setProgressBackgroundColorSchemeColor(ThemeManager.PRIMARY)

        val manager = LinearLayoutManager(activity, RecyclerView.VERTICAL, false)
        list.setHasFixedSize(true)
        list.layoutManager = manager

        refreshLayout.isEnabled = false

        getCachedUsers(0, 0)
        if (savedInstanceState == null)
            loadUsers(FRIENDS_COUNT, 0)

        if (adapter != null && list?.adapter == null) {
            list?.adapter = adapter
        }
    }

    override fun getCachedUsers(count: Int, offset: Int) {
        val users = getFriends(UserConfig.userId, onlyOnline)

        if (!ArrayUtil.isEmpty(users)) {
            setRefreshing(true)
            setNoItemsViewVisible(false)
            createAdapter(users, 0)
        } else {
            setProgressBarVisible(true)
        }
    }

    override fun loadUsers(count: Int, offset: Int) {
        if (!Util.hasConnection()) {
            showNoInternetToast()
            return
        }

        setRefreshing(true)

        TaskManager.execute {

            VKApi.friends().get().userId(UserConfig.userId).order("hints")
                .fields(VKUser.FIELDS_DEFAULT)
                .execute(VKUser::class.java, object : OnResponseListener {
                    override fun onComplete(models: ArrayList<*>?) {
                        if (ArrayUtil.isEmpty(models)) return
                        models ?: return

                        val users = models as ArrayList<VKUser>

                        if (offset == 0) {
                            CacheStorage.delete(DatabaseHelper.FRIENDS_TABLE)
                            CacheStorage.insert(DatabaseHelper.FRIENDS_TABLE, users)
                        }

                        CacheStorage.insert(DatabaseHelper.USERS_TABLE, users)

                        createAdapter(sortUsers(users), offset)

                        presenter.onFilledList()
                    }

                    override fun onError(e: Exception) {
                        setRefreshing(false)
                        showErrorToast()
                    }
                })

            refreshLayout.isRefreshing = false
        }
    }

    private fun sortUsers(users: ArrayList<VKUser>): ArrayList<VKUser> {
        if (!onlyOnline) return users
        val newUsers = arrayListOf<VKUser>()

        for (user in users) {
            if ((onlyOnline && user.isOnline)) {
                newUsers.add(user)
            }
        }

        return newUsers
    }

    override fun createAdapter(items: ArrayList<VKUser>?, offset: Int) {
        if (ArrayUtil.isEmpty(items)) return

        if (adapter == null) {
            adapter = UserAdapter(this, items!!)
            list!!.adapter = adapter
            return
        }

        if (offset != 0) {
            adapter!!.values!!.addAll(items!!)
            adapter!!.notifyDataSetChanged()
            return
        }

        adapter!!.changeItems(items!!)
        adapter!!.notifyDataSetChanged()
    }


    override fun showNoInternetToast() {
        setRefreshing(false)
        Toast.makeText(activity, R.string.connect_to_the_internet, Toast.LENGTH_SHORT).show()
    }

    override fun showErrorToast() {
        Toast.makeText(activity, getString(R.string.error), Toast.LENGTH_LONG).show()
    }

    override fun setProgressBarVisible(visible: Boolean) {
        when (visible) {
            true -> {
                progressBar.visibility = View.VISIBLE
                setRefreshing(false)
                setNoItemsViewVisible(false)
            }
            else -> progressBar.visibility = View.GONE
        }
    }

    override fun setRefreshing(value: Boolean) {
        when (value) {
            true -> {
                refreshLayout.isRefreshing = true
                setProgressBarVisible(false)
            }
            else -> refreshLayout.isRefreshing = false
        }
    }

    override fun setNoItemsViewVisible(visible: Boolean) {
        when (visible) {
            true -> {
                emptyView.visibility = View.VISIBLE
                setProgressBarVisible(false)
            }
            else -> emptyView.visibility = View.GONE
        }
    }

    override fun clearList() {
        adapter ?: return
        adapter!!.clear()
        adapter!!.notifyDataSetChanged()
    }

    fun openChat(position: Int) {
        val user = adapter!!.getItem(position)

        val args = Bundle().apply {
            putString("title", user.toString())
            putString("photo", user.photo200)
            putInt("peer_id", user.id)
        }

        val canWrite = !user.isDeactivated

        args.putBoolean("can_write", canWrite)

        if (!canWrite) {
            args.putInt("reason", VKConversation.getReason(VKConversation.Reason.USER_DELETED))
        }

        FragmentSelector.selectFragment(fragmentManager!!, FragmentMessages(), args, true)
    }

    fun showDialog(position: Int, v: View) {
        val menu = PopupMenu(activity!!, v)
        menu.inflate(R.menu.fragment_friends_funcs)
        menu.setOnMenuItemClickListener {
            showConfirmDeleteFriend(position)
            true
        }
        menu.show()
    }

    private fun showConfirmDeleteFriend(position: Int) {
        val adb = AlertDialog.Builder(activity!!)
        adb.setTitle(R.string.confirmation)
        adb.setMessage(R.string.confirm_delete_friend)
        adb.setPositiveButton(R.string.yes) { _, _ -> deleteFriend(position) }
        adb.setNegativeButton(R.string.no, null)
        adb.show()
    }

    private fun deleteFriend(position: Int) {
        if (!Util.hasConnection()) {
            refreshLayout.isRefreshing = false
            return
        }

        refreshLayout.isRefreshing = true

        val user = adapter!!.getItem(position)
        val userId = user.id

        TaskManager.execute {
            VKApi.friends().delete().userId(userId).execute(null, object : OnResponseListener {
                override fun onComplete(models: ArrayList<*>?) {
                    adapter!!.remove(position)
                    adapter!!.notifyItemRemoved(position)
                    adapter!!.notifyItemRangeChanged(0, adapter!!.itemCount, -1)
                    refreshLayout.isRefreshing = false

                    CacheStorage.delete(
                        DatabaseHelper.FRIENDS_TABLE,
                        DatabaseHelper.USER_ID,
                        userId
                    )
                }

                override fun onError(e: Exception) {
                    Log.e("Error delete friend", Log.getStackTraceString(e))
                    refreshLayout.isRefreshing = false
                    Toast.makeText(activity, R.string.error, Toast.LENGTH_SHORT).show()
                }
            })
        }
    }

    fun onReceive(data: Array<Any>) {
        adapter ?: return
        when (data[0] as String) {
            Keys.USER_OFFLINE -> {
                val index = adapter!!.searchUser(data[1] as Int)
                if (index != -1) {
                    adapter!!.remove(index)
                    adapter!!.notifyDataSetChanged()
                }
            }
            Keys.USER_ONLINE -> {
                val user = MemoryCache.getUser(data[1] as Int)
                if (user != null) {
                    adapter!!.add(0, user)
                    adapter!!.notifyDataSetChanged()
                }
            }
        }
    }

    companion object {
        private const val FRIENDS_COUNT = 30
    }
}